File: cpp/leetcode/212.cpp
class TrieNode {
public:
    TrieNode *child[26];
    string str;
    TrieNode(){
        str = "";
        for (auto &ptr : child) ptr = NULL;
    }
};

class Trie {
public:
    TrieNode *root;
    Trie(){
        root = new TrieNode();
    }
    void insert(string word) {
        TrieNode *ptr = root;
        for (auto ch : word) {
            int i = ch - 'a';
            if (!ptr->child[i]) ptr->child[i] = new TrieNode();
            ptr = ptr->child[i];
        }
        ptr->str = word;
    }
};

class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        vector<string> res;
        if (words.empty() || board.empty() || board[0].empty()) return res;
        int m = board.size(), n = board[0].size();
        vector<vector<bool>> visit(m, vector<bool>(n, false));
        Trie T;
        for (auto word : words) T.insert(word);
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j){
                int idx = board[i][j] - 'a';
                if (T.root->child[idx]){ 
                    search(board, T.root->child[idx], i, j, visit, res);
                }
            }
        }
        return res;
    }

    void search(vector<vector<char>> &board, TrieNode *p, int i, int j, vector<vector<bool>> &visit, vector<string> &res) { 
        if (!p->str.empty()) {
            res.push_back(p->str);
            p->str.clear();
        }
        int d[][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        int m = board.size(), n = board[0].size();
        visit[i][j] = true;
        for (auto &a : d) {
            int nx = a[0] + i, ny = a[1] + j;
            if (nx >= 0 && nx < m && ny >= 0 && ny < n && !visit[nx][ny] && p->child[board[nx][ny] - 'a']) 
                search(board, p->child[board[nx][ny] - 'a'], nx, ny, visit, res);
        }
        visit[i][j] = false;
    }
};

File: python/leetcode/212.py
class Solution(object):
    def findWords(self, board, words):
        root = {}
        for word in words:
            node = root
            for ch in word:
                if ch not in node:
                    node[ch] = {}
                node = node[ch]  # child
            node['#'] = '#'  # end
        res = []
        for i in range(len(board)):
            for j in range(len(board[0])):
                self.find(board, i, j, root, '', res)
        return list(set(res))

    def find(self, board, i, j, root, path, res):
        if '#' in root: res.append(path)
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or board[i][j] not in root:
            return
        tmp, board[i][j] = board[i][j], "@"
        self.find(board, i + 1, j, root[tmp], path + tmp, res)
        self.find(board, i, j + 1, root[tmp], path + tmp, res)
        self.find(board, i - 1, j, root[tmp], path + tmp, res)
        self.find(board, i, j - 1, root[tmp], path + tmp, res)
        board[i][j] = tmp